ledger/accountdb_test.go

```diff
@@ -1282,205 +1282,221 @@ func TestCompactResourceDeltas(t *testing.T) {
-	rd.SetAppLocalState(appState)
-	rd.SetAppParams(appParams, true)
-	a.True(rd.IsApp())
-	a.True(rd.IsOwning())
-	a.True(rd.IsHolding())
-	a.False(rd.IsEmptyAppFields())
-	a.False(rd.IsEmpty())
-
-	// 3. both exist: holding not empty, param is empty
-	rd = makeResourcesData(0)
-	rd.SetAppLocalState(appState)
-	rd.SetAppParams(appParamsEmpty, true)
-	a.True(rd.IsApp())
-	a.True(rd.IsOwning())
-	a.True(rd.IsHolding())
-	a.False(rd.IsEmptyAppFields())
-	a.False(rd.IsEmpty())
-
-	// 4. both exist: holding empty, param is not empty
-	rd = makeResourcesData(0)
-	rd.SetAppLocalState(appLocalEmpty)
-	rd.SetAppParams(appParams, true)
-	a.True(rd.IsApp())
-	a.True(rd.IsOwning())
-	a.True(rd.IsHolding())
-	a.False(rd.IsEmptyAppFields())
-	a.False(rd.IsEmpty())
-
-	// 5. holding does not exist and params is empty
-	rd = makeResourcesData(0)
-	rd.SetAppParams(appParamsEmpty, false)
-	a.True(rd.IsApp())
-	a.True(rd.IsOwning())
-	a.False(rd.IsHolding())
-	a.True(rd.IsEmptyAppFields())
-	a.False(rd.IsEmpty())
-
-	// 6. holding does not exist and params is not empty
-	rd = makeResourcesData(0)
-	rd.SetAppParams(appParams, false)
-	a.True(rd.IsApp())
-	a.True(rd.IsOwning())
-	a.False(rd.IsHolding())
-	a.False(rd.IsEmptyAppFields())
-	a.False(rd.IsEmpty())
-
-	// 7. holding exist and not empty and params does not exist
-	rd = makeResourcesData(0)
-	rd.SetAppLocalState(appState)
-	a.True(rd.IsApp())
-	a.False(rd.IsOwning())
-	a.True(rd.IsHolding())
-	a.False(rd.IsEmptyAppFields())
-	a.False(rd.IsEmpty())
-
-	// 8. both do not exist
-	rd = makeResourcesData(0)
-	a.False(rd.IsApp())
-	a.False(rd.IsOwning())
-	a.False(rd.IsHolding())
-	a.True(rd.IsEmptyAppFields())
-	a.True(rd.IsEmpty())
+	// Since some steps use randomly generated input, the test is run N times
+	// to cover a larger search space of inputs.
+	for i := 0; i < 1000; i++ {
+		rd := resourcesData{}
+		a.False(rd.IsApp())
+		a.True(rd.IsEmpty())
+
+		rd = makeResourcesData(1)
+		a.False(rd.IsApp())
+		a.False(rd.IsHolding())
+		a.False(rd.IsOwning())
+		a.True(rd.IsEmpty())
+
+		// check empty
+		appParamsEmpty := basics.AppParams{}
+		rd = resourcesData{}
+		rd.SetAppParams(appParamsEmpty, false)
+		a.True(rd.IsApp())
+		a.True(rd.IsOwning())
+		a.True(rd.IsEmptyAppFields())
+		a.False(rd.IsEmpty())
+		a.Equal(appParamsEmpty, rd.GetAppParams())
+
+		appLocalEmpty := basics.AppLocalState{}
+		rd = resourcesData{}
+		rd.SetAppLocalState(appLocalEmpty)
+		a.True(rd.IsApp())
+		a.True(rd.IsHolding())
+		a.True(rd.IsEmptyAppFields())
+		a.False(rd.IsEmpty())
+		a.Equal(appLocalEmpty, rd.GetAppLocalState())
+
+		// check both empty
+		rd = resourcesData{}
+		rd.SetAppLocalState(appLocalEmpty)
+		rd.SetAppParams(appParamsEmpty, true)
+		a.True(rd.IsApp())
+		a.True(rd.IsOwning())
+		a.True(rd.IsHolding())
+		a.True(rd.IsEmptyAppFields())
+		a.False(rd.IsEmpty())
+		a.Equal(appParamsEmpty, rd.GetAppParams())
+		a.Equal(appLocalEmpty, rd.GetAppLocalState())
+
+		// check empty states + non-empty params
```

Comment:
Maybe move the start of the for-loop down to here?  The above is all deterministic, right?

Reply:
@jannotti Responding inline below:

> I found the changes by looking only at the inserted if statements,

Correct.

> The above is all deterministic, right?

Correct - Though it is liable to change in the future.

>  move the start of the for-loop down to here? 

I don't think it materially improves the situation, but I'm not opposed.  Change made in  https://github.com/algorand/go-algorand/pull/4699/commits/18f34b5f7d9886329033440f978497d988b9928a.